sycl/source/detail/error_handling/enqueue.cpp

```diff
@@ -0,0 +1,179 @@
+    if (!HasLocalSize && (Ver[0] == '1' || (Ver[0] == '2' && Ver[2] == '0'))) {
+      throw sycl::nd_range_error(
+          "OpenCL 1.x and 2.0 requires to pass local size argument even if "
+          "required work-group size was specified in the program source",
+          PI_INVALID_WORK_GROUP_SIZE);
+    }
+  }
+
+  if (Ver[0] == '1') {
+    // OpenCL 1.x:
+    // CL_INVALID_WORK_GROUP_SIZE if local_work_size is specified and the
+    // total number of work-items in the work-group computed as
+    // local_work_size[0] * ... * local_work_size[work_dim – 1] is greater
+    // than the value specified by CL_DEVICE_MAX_WORK_GROUP_SIZE in
+    // table 4.3
+    size_t MaxWGSize;
+    PI_CALL(RT::piDeviceGetInfo, Device, PI_DEVICE_MAX_WORK_GROUP_SIZE,
+            sizeof(size_t), &MaxWGSize, nullptr);
+    if (NDRDesc.LocalSize[0] * NDRDesc.LocalSize[1] * NDRDesc.LocalSize[2] >
+        MaxWGSize) {
+      throw sycl::nd_range_error(
+          "Total number of work-items in a work-group cannot exceed "
+          "info::kernel_work_group::work_group_size which is equal to " +
+              std::to_string(MaxWGSize),
+          PI_INVALID_WORK_GROUP_SIZE);
+    }
+  } else {
+    // OpenCL 2.x:
+    // CL_INVALID_WORK_GROUP_SIZE if local_work_size is specified and the
+    // total number of work-items in the work-group computed as
+    // local_work_size[0] * ... * local_work_size[work_dim – 1] is greater
+    // than the value specified by CL_KERNEL_WORK_GROUP_SIZE in table 5.21.
+    size_t KernelWGSize;
+    PI_CALL(RT::piKernelGetGroupInfo, Kernel, Device, CL_KERNEL_WORK_GROUP_SIZE,
+            sizeof(size_t), &KernelWGSize, nullptr);
+    if (NDRDesc.LocalSize[0] * NDRDesc.LocalSize[1] * NDRDesc.LocalSize[2] >
+        KernelWGSize) {
+      throw sycl::nd_range_error(
+          "Total number of work-items in a work-group cannot exceed "
+          "info::device::max_work_group_size which is equal to " +
+              std::to_string(KernelWGSize) + " for this kernel",
+          PI_INVALID_WORK_GROUP_SIZE);
+    }
+  }
+
+  if (HasLocalSize) {
+    bool NonUniformWGs = (NDRDesc.LocalSize[0] != 0 &&
+                          NDRDesc.GlobalSize[0] % NDRDesc.LocalSize[0] != 0) ||
+                         (NDRDesc.LocalSize[1] != 0 &&
+                          NDRDesc.GlobalSize[1] % NDRDesc.LocalSize[1] != 0) ||
+                         (NDRDesc.LocalSize[2] != 0 &&
+                          NDRDesc.GlobalSize[2] % NDRDesc.LocalSize[2] != 0);
+
+    if (Ver[0] == '1') {
+      // OpenCL 1.x:
+      // CL_INVALID_WORK_GROUP_SIZE if local_work_size is specified and
+      // number of workitems specified by global_work_size is not evenly
+      // divisible by size of work-group given by local_work_size
+
+      if (NonUniformWGs)
+        throw sycl::nd_range_error(
+            "Non-uniform work-groups are not allowed. Moreover, underlying "
+            "OpenCL 1.x doesn't support non-uniform work-group sizes",
```

Comment:
Non-uniform work-groups are not supported by the target device.

Reply:
Done